proposals/attach-to-property.md

```diff
@@ -0,0 +1,142 @@
+- "attachTo": placeholder for the name of the operator function
+- "constructor" means either primary constructor, or each of secondary constructor if there is no primary constructor
+
+## Motivation / use cases
+
+Make delegate instance aware of host instance and property it was used with.
+
+- Register host instance in external table of managed entities, e.g. database schema, component container.
+- Register delegate properties in host instance with callback for "metadata", e.g. for database columns.
+- Call back into host instance with additional information, e.g. property metadata, delegate instance.
+  Can be used to communicate delegate instance back to the host instance for later use, by declaring member
+  extension `attachTo` function with receiver of delegate type.
+- Execute custom code before first access to property, e.g. `async` property that is being calculated
+  in the background, and blocks as needed when trying to use the value. With coroutines, we can suspend instead
+  of blocking.
+
+## Description
+
+Client code:
+```
+class HostType {
+   val name by DelegateType()
+}
+```
+
+Attach operator:
+```
+operator fun DelegateType.attachTo(instance: HostType, property: KProperty<*>) : Unit`
+```
+
+- No extra instances are created, all information is already available
+
+## Alternative names
+
+- bindToInstance
+- delegatedFrom
+
+## Implementation
+
+### Syntax
+
+No new syntax is required.
+
+### Resolution
+
+- resolving the `attachTo` function happens similar to resolving `getValue`/`setValue` functions.
+- if there are more than one `attachTo` functions available, overload resolution selects best candidate.
+- `attachTo` that is not accessible is ignored.
+- `attachTo` can be inline.
+- `attachTo` can be generic, inference happens normally based on the type of host instance.
+- Member extension `attachTo` on matching delegate type is preferred.
+- No implicit `attachTo` operator from Java
+
+### Diagnostics
+
+Declaration site:
+- `attachTo` should have Unit return type
+- `attachTo` should have two parameters, second one should have `KProperty<*>` type
+
+Use site (`by` keyword):
+-
+
+### Code generation (JVM)
+
+At the end of the primary constructor, or each of secondary constructors:
+For each property with discovered `attachTo` operator:
+Call to `attachTo` is generated passing `this` and property metadata.
+
+Member `attachTo`
+```
+name$delegate.attachTo(this, $$delegatedProperties[0])
+```
+
+Extension `attachTo`
+```
+FileKt.attachTo(name$delegate, this, $$delegatedProperties[0])
+```
+
+Member Extension `attachTo`
+```
+this.attachTo(name$delegate, this, $$delegatedProperties[0])
+```
+
+Notes:
+- inline `attachTo` function is inlined normally in place of call
+
+### Code generation (JS)
+
+Similar to JVM.
+
+### Debug information
+
+- Breakpoint on a property should be triggered on `attachTo` call.
+- Step over on a property should go to next property with `attachTo` generated
```

Comment:
Since these calls are generated at the end of constructor, there is nothing that follows these calls. Until another feature generates something at the end of constructor, of course. Alternative would be generate `attachTo` calls next to PUTFIELD, but that means attachment handlers will see incomplete host instance.